GD/__init__.py

```diff
@@ -11,16 +15,203 @@
+        sorts = ["date asc", "avg_rating desc",
+                 "num_favorites desc", "downloads desc"]
+        self.parms = {'debug': 'false',
+                      'xvar': 'production',
+                      'total_only': 'false',
+                      'count': '10000',
+                      'sorts': ",".join(sorts),
+                      'fields': ",".join(fields)}
+
+    @abc.abstractmethod
+    def get_tapes(self, years):
+        """Get a list of tapes."""
+        pass
+
+
+class TapeDownloader(BaseTapeDownloader):
+    """Synchronous Grateful Dead Tape Downloader"""
+
+    def get_tapes(self, years):
+        """Get a list of tapes.
+
+        Parameters:
+
+            years: List of years to download tapes for
+
+        Returns a list dictionaries of tape information
+        """
+        tapes = []
+
+        for year in years:
+            year_tapes = self._get_tapes_year(year)
+            tapes.extend(year_tapes)
+
+        return tapes
+
+    def _get_tapes_year(self, year):
+        """Get tape information for a year.
+
+        Parameters:
+
+            year: The year to download tape information for
+
+        Returns a list of dictionaries of tape information
+        """
+        current_rows = 0
+        tapes = []
+        r = self._get_chunk(year)
+        j = r.json()
+        total = j['total']
+        logger.debug(f"total rows {total}")
+        current_rows += j['count']
+        tapes = j['items']
+        while current_rows < total:
+            cursor = j['cursor']
+            r = self._get_chunk(year, cursor)
+            j = r.json()
+            cursor = j['cursor']
+            current_rows += j['count']
+            tapes.extend(j['items'])
+        return tapes
+
+    def _get_chunk(self, year, cursor=None):
+        """Get one chunk of a year's tape information.
+
+        Parameters:
+
+            year: The year to download tape information for
+            cursor: Used to download a segment of a year of tapes
+
+        Returns a list of dictionaries of tape information
+        """
+        parms = self.parms.copy()
+        if cursor is not None:
+            parms['cursor'] = cursor
+        query = 'collection:GratefulDead AND year:'+str(year)
+        parms['q'] = query
+        r = requests.get(self.api, params=parms)
+        logger.debug(f"url is {r.url}")
+        if r.status_code != 200:
+            logger.error(f"Error {r.status_code} collecting data")
+            raise Exception(
+                'Download', 'Error {} collection'.format(r.status_code))
+        return r
+
+
+class AsyncTapeDownloader(BaseTapeDownloader):
+    """Asynchronous Grateful Dead Tape Downloader"""
+
+    def get_tapes(self, years):
+        """Get a list of tapes.
+
+        Parameters:
+
+            years: List of years to download tapes for
+
+        Returns a list dictionaries of tape information
+        """
+        tapes = asyncio.run(self._get_tapes(years))
```

Comment:
I believe this line requires python 3.7. There is an alternate syntax for older python.